// @fitbit/bison-i18n/mod/LanguageTable.js
import { __read, __values } from "tslib";
import pofile from 'pofile';
import { TextEncoder } from 'text-encoding-utf-8';
import fbpubfletcher32 from './fbpubfletcher32';
var HASH_MASK = 0x3fffffff;
function hashUTF8(string) {
    return fbpubfletcher32(new TextEncoder('utf-8').encode(string).buffer) & HASH_MASK;
}
var _Message = (function () {
    function _Message(msgid) {
        this.msgid = msgid;
        this.msgstr = {};
        this.maxLength = 0;
    }
    _Message.prototype.addTranslation = function (language, string) {
        var utf8String = new TextEncoder('utf-8').encode(string);
        this.msgstr[language] = utf8String;
        this.maxLength = Math.max(this.maxLength, utf8String.length);
    };
    return _Message;
}());
export { _Message };
var TranslationLoader = (function () {
    function TranslationLoader() {
        this.languages = new Set();
        this.messages = new Map();
    }
    TranslationLoader.prototype.loadLanguage = function (language, podata) {
        var e_1, _a;
        if (this.languages.has(language)) {
            throw new Error("Language '" + language + "' already loaded");
        }
        var po = pofile.parse(podata);
        try {
            for (var _b = __values(po.items), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = _c.value, msgid = _d.msgid, msgstr = _d.msgstr;
                if (msgstr.length > 1) {
                    throw new Error("msgid \"" + msgid + "\" has multiple msgstr values. This is not supported.");
                }
                var hashKey = hashUTF8(msgid);
                var message = this.messages.get(hashKey);
                if (message === undefined) {
                    message = new _Message(msgid);
                    this.messages.set(hashKey, message);
                }
                if (message.msgid !== msgid) {
                    throw new Error("msgid \"" + msgid + "\" hashes to the same value as \"" + message.msgid + "\" -- tweak msgid and try again");
                }
                message.addTranslation(language, msgstr[0]);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.languages.add(language);
        return this;
    };
    TranslationLoader.prototype.build = function () {
        return new LanguageTable(this.languages, this.messages);
    };
    return TranslationLoader;
}());
export { TranslationLoader };
var LanguageTable = (function () {
    function LanguageTable(languages, messages) {
        this.languages = languages;
        this.messages = messages;
    }
    LanguageTable.prototype.getLanguage = function (language) {
        var e_2, _a, e_3, _b;
        var ids = [];
        var stringTableLength = 0;
        try {
            for (var _c = __values(this.messages.entries()), _d = _c.next(); !_d.done; _d = _c.next()) {
                var _e = __read(_d.value, 2), id = _e[0], message = _e[1];
                var msgstr = message.msgstr[language];
                if (msgstr !== undefined) {
                    ids.push(id);
                    stringTableLength += msgstr.byteLength + 1;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (ids.length === 0)
            throw new Error("There are no translation strings for \"" + language + "\"");
        ids.sort(function (a, b) { return a - b; });
        var headerLength = 4
            + 8 * ids.length
            + 2;
        var outArray = new Uint8Array(headerLength + stringTableLength);
        var dataView = new DataView(outArray.buffer);
        var headerCursor = 0;
        var stringOffset = 0;
        dataView.setUint16(headerCursor, ids.length, true);
        headerCursor += 4;
        try {
            for (var ids_1 = __values(ids), ids_1_1 = ids_1.next(); !ids_1_1.done; ids_1_1 = ids_1.next()) {
                var id = ids_1_1.value;
                var message = this.messages.get(id);
                dataView.setUint32(headerCursor, id, true);
                headerCursor += 4;
                dataView.setUint16(headerCursor, stringOffset, true);
                headerCursor += 2;
                dataView.setUint16(headerCursor, message.maxLength, true);
                headerCursor += 2;
                var msgstr = message.msgstr[language];
                outArray.set(msgstr, headerLength + stringOffset);
                stringOffset += msgstr.byteLength + 1;
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (ids_1_1 && !ids_1_1.done && (_b = ids_1.return)) _b.call(ids_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        dataView.setUint16(headerCursor, stringOffset, true);
        return outArray.buffer;
    };
    return LanguageTable;
}());
export { LanguageTable };
